url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/c64utils/Assembler.java

```diff
@@ -0,0 +1,694 @@
+package com.dreamfabric.c64utils;
+
+import java.util.Hashtable;
+import java.util.Vector;
+import com.dreamfabric.jac64.*;
+import java.io.InputStream;
+
+/**
+ * A simple assembler for JaC64 (C64) / 6502
+ * Joakim Eriksson, JaC64.com / Dreamfabric.com
+ * ----------------------------------------------
+ * Should be almost compatible (?) with a65 and
+ * some other C64 assemblers
+ * ----------------------------------------------
+ * To get full functionality for loading binaries
+ * you need to override openBinary!
+ *
+ * Created: Fri Dec 29 09:52:38 2006
+ * 
+ * @author Joakim Eriksson
+ * @version 1.0
+ */
+public class Assembler {
+
+  public static final int DEBUG_LEVEL = 0;
+
+  public static final int CODE = 1;
+  public static final int COMMENT = 2;
+
+  public static final int REF_WORD = 0;
+  public static final int REF_BYTE = 1;
+  public static final int REF_BYTE_LO = 2;
+  public static final int REF_BYTE_HI = 3;
+  public static final int REF_RELATIVE = 4;
+
+  public static final int[] REF_SIZE = new int[] {
+    2,1,1,1,1
+  };
+  private int mode = CODE;
+  // pos => address
+  private int pos = 0;
+  private int lastPos = 0;
+
+  private Hashtable labels = new Hashtable();
+
+  private Vector references = new Vector();
+
+  private String currentLine;
+  private String[] tokens;
+  private int[] memory = new int[0x10000];
+  private int lineNo;
+  protected String workingDir;
+
+
+  /**
+   * Creates a new <code>Assembler</code> instance.
+   *
+   */
+  public Assembler() {
+  }
+
+  public void setMemory(int[] memory) {
+    this.memory = memory;
+  }
+
+  public void setWorkingDir(String dir) {
+    workingDir = dir;
+    if (!workingDir.endsWith("/")) {
+      workingDir += "/";
+    }
+    System.out.println("Set working dir to " + workingDir);
+  }
+
+  public int[] assemble(String s, int start) {
+    String[] lines = split(s, "\n", false);
+    labels.clear();
+    references.removeAllElements();
+
+    setPos(start);
+    mode = CODE;
+    // First pass...
+    for (int i = 0, n = lines.length; i < n; i++) {
+      currentLine = lines[i];
+      if (assembleLine(lines[i], memory, pos) == 1) break;
+      lineNo++;
+    }
+
+    resolve();
+
+    return memory;
+  }
+
+  private void setPos(int start) {
+    pos = start;
+    lastPos = pos;
+    System.out.println("Location set to: " + start);
+  }
+
+  private int assembleLine(String lineOrig, int[] memory, int adr) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 155 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 101-300 into a separate method.